### ISSUE TITLE
long double on x86 32bits should compile with -mfpmath=sse

### ISSUE THREAD

**Issue by fredericopissarra** [firehol/netdata#263]
Why do you use _long double_ type in storage_number.{c,h}? Do you need more than 15 digits of precision? Or this reencoding of floating point values is made just to avoid using {f,s}]printf() routines?

If the values are reencoded to use 24 bits of precision, _float_ should be enough (but _double_ is a little bit faster)... And, to use _long double_ is to left aside SSE and rely on fp87 instructions (slower)...

Where did I get this 24 bits precision? From the comments on pack_storage_number() function:

`// bit 32 = sign 0:positive, 1:negative`
`// bit 31 = 0:divide, 1:multiply`
`// bit 30, 29, 28 = (multiplier or divider) 0-6 (7 total)`
`// bit 27, 26, 25 flags`
`// bit 24 to bit 1 = the value`

Are there any reason for this decision?

Maybe it is better to use something like fpconv [fpconv](https://github.com/night-shift/fpconv) and store the floating point values as _double_ (53 bits of precision).

[]s
Fred

**Comment by ktsaou** [comment]
well, I needed a number that can fit a wider range of possible numbers, with some detail loss. So, it rounds the integer part too, but no more than 0.0001%.

The result is that I can fit in 29 bits, numbers from `-167772150000000.0` to `167772150000000.0` and decimal numbers with detail up to `0.00001`.

Of course bigger number have less detail (even the integer part of them).

So, I can have `0.00001` or `167772150000000.0` stored in 29 bits.

**Comment by fredericopissarra** [comment]
Thanks...

The accuracy of 0.00001 and the precision of 8 decimal algarisms (16777215.0*10⁷) can easily be done with _float_. 

Using _double_ you'll have 15 decimal algarisms precision (167772151234567.0 could be stored 'precisely') and alse detail up to 1e-15, with no floating point re-encoding... It will only take 64 bits and it is supported by SSE2. _long double_ is slow...

I'm testing fpconv and it is almost 6 times faster than gcvt() [or sprintf]...

[]s
Fred

**Comment by ktsaou** [comment]
ok. Regarding the flags:

I need to store everything in the time series database in 32 bits. So the whole database is just an array on X 32 bit numbers.

These are needed:

1. the number itself
2. if a number has been collected or not for this point int time
3. if the collected number (not the stored) has overflown or not
4. if there are related annotations to the collected number (not currently used)

Of course I could use just one bit for all flags and have a lookup somewhere else to find if the number is present and if it is overflown. But this would make every chart refresh a lot more complex, so I decided that just the annotations should be outside the number itself.

Regarding the calculations, I use long double because the collected values are in many cases `unsigned long long int` and netdata interpolates them to second boundaries, using nanosecond detail (after converting them using the supplied multiplier and divisor to change their units). So, with float it would overflow.

**Comment by ktsaou** [comment]
You can change the `typedef` of `calculated_number` to something else and recompile netdata to see it in action. This is why I typedef'd them.

**Comment by ktsaou** [comment]
I think we don't need this open. You already work on alternatives.

**Comment by fredericopissarra** [comment]
There is another interesting "bug" on using 387 instructions (the compiler will be forced to do it if 'long double' is used) called "double rounding". When a value is loaded to a 387 register it is rounded to 'long double' converted from a 'double' value. Operations rounds the results again... In i386 mode (32 bits OS), the GCC default to use 387. In x86_64 mode it defaults to SSE2.

On x86_64, even if you use 'long double', sometimes SSE instructions are used to deal with intermediary values. Then, x86_64 suffers from double rounding as well...

387 will, always, do its magic using extended precision...

I think it is prudent to add the option -mfpmath=sse when compiling for 32 bits to avoid this bug...

Here a small code for a test (using 'double'): https://ghostbin.com/paste/zeuuy

And, if the environment allows FMA instructions, it will avoid double rounding of other kind (expressions like 'a*b+c' rounds twice because we have 2 operations.

[]s
Fred

**Comment by ktsaou** [comment]
nice!

> I think it is prudent to add the option -mfpmath=sse when compiling for 32 bits to avoid this bug...

ok. this has to be done only on 32bit x86 machines. Other architectures should not have it.

@alonbl @philwhineray any ideas how the automake system can detect the machine is x86 32bit and enable this flag?

**Comment by fredericopissarra** [comment]
In one of my projects I use this in the Makefile:

   ARCH = $(shell arch)
   ifneq ($(ARCH),x86_64)
     CFLAGS += -msse2 -mfpmath=sse
   endif

But this will work only on x86 platforms. I don't have a 32 bit OS 
installed here, so I think you'll have to find the correct string do 
match arch,

**Comment by glensc** [comment]
that is not portable, also prevents cross-compilation.

you should rather do TRY_COMPILE with specified `$CC` and then append `$CFLAGS`

**Comment by fredericopissarra** [comment]
Maybe this is useful too:

$ gcc -dumpmachine

**Comment by glensc** [comment]
[AC_TRY_COMPILE](http://ftp.gnu.org/old-gnu/Manuals/autoconf-2.13/html_node/autoconf_42.html) manual

**Comment by philwhineray** [comment]
This presumably wraps the same and will take care of the test and also
adding the flag if it works:

http://www.gnu.org/software/autoconf-archive/ax_check_compile_flag.html

**Comment by ktsaou** [comment]
> optimizations and workarounds should probably not be added to a package.

Probably we can handle this at the installer, not the makefiles.

The installer already passes `CFLAGS="-O3"`, so we could add a check there and pass the proper options to the compiler.

So `$((arch))` should return `i386` and `i686` for 32bit x86?

**Comment by alonbl** [comment]
Again, I am unsure what installer is.
The packager (rpm maintainer, ebuild maintainer, deb maintainer etc...) is responsible of providing/customizing CFLAGs for optimization cases, if required.

**Comment by glensc** [comment]
again, **do not use** `arch`! it's not portable and gives bogus results when cross compiling or for example running i686 root on x86-86 kernel!

**Comment by simonnagl** [comment]
I have the same opinion like @alonbl and @glensc.

Package maintainer should be able to select `CFLAG` for optimization purposes. We should not include them into the "manual" build tools like `configure` and `make`

Package maintainers should not use `netdata-installer.sh`. The Use Case of it is to have a simple method, so everybody can build and install `netdata` easily. (Am I right?)
There auto-detection of optimization may be possible.

Do we want that cross compiling is available with `netdata-installer.sh`?

If yes we need a way to specific for what system we want to compile. Then we can automate optimization for specified hardware. If nothing is specified the installer can autodetect the system.

**Comment by simonnagl** [comment]
For me this sounds good.

**Comment by ktsaou** [comment]
> Sane defaults should not include optimizations, this belongs to
> documentation.

Sorry. I strongly disagree. The world does not work that way. Never did. Never will.

**Comment by ktsaou** [comment]
PR #446
Please check it.

**Comment by alonbl** [comment]
On 23 May 2016 at 21:29, Costa Tsaousis <notifications@github.com> wrote:

> Sane defaults should not include optimizations, this belongs to
> documentation.
>
> Sorry. I strongly disagree. The world does not work that way. Never did.
> Never will.
>

It has always worked this way, look at all well behaved packages out there.
You just did not notice because you use packager's output.

**Comment by alonbl** [comment]
Do whatever you want in the "install" script, it should never be used
anyway.
It is a bad practice to provide such.
Sysadmin (or user) should use their favorite packager tool to install
components to keep system integrity.
This is why we have spec, ebuild, etc...

Anyway, please do not touch (for this purpose) autotools/build that is used
for production (packaging).

**Comment by ktsaou** [comment]
Don't be afraid. The autotools/build system will remain packager friendly.

**Comment by glensc** [comment]
https://github.com/firehol/netdata/issues/263#issuecomment-221039543

i don't agree here, i think package should have configure option to detect/enable/disable sse/sse2 and use [AX_CHECK_COMPILE](https://github.com/firehol/netdata/issues/263#issuecomment-220958036) for availability:

- if option is enabled and not available -> error, exit
- if option is disabled, no check is attempted -> no sse2 flags added
- if option is not specified, it is enabled if compile tests succeeds

this gives most flexibility for packagers and for distributors.

**Comment by fredericopissarra** [comment]
Instead of using arch you can always do something like this:

```
#!/bin/bash

OS="`uname -s`"
MACHINE="`uname -m`"
if [ "$MACHINE" == "i386" ]; then
  case $OS in
    Linux)
      if grep sse2 /proc/cpuinfo; then
        CFLAGS="${CFLAGS} -msse2 -mfpmath=sse"
      fi
      ;;
    FreeBSD)
      if grep SSE2 /var/run/dmesg.boot; then
        CFLAGS="${CFLAGS} -msse2 -mfpmath=sse"
      fi
      ;;
    ...
  esac
fi
```

**Comment by ktsaou** [comment]
closed since PR #1032 adds the required functionality (on x86 `-msse2 -mfpmath=sse` is added automatically if the compiler accepts it).
